### PR TITLE
Refactor contextmodule to es6

### PR DESCRIPTION
**Pull request by timse**
**What kind of change does this PR introduce?**

refactor context module to esnext

**Did you add tests for your changes?**

N/A

**If relevant, link to documentation update:**

N/A

**Summary**

refactor context module to esnext and split up in smaller methods to make it easier to reason about.

**Does this PR introduce a breaking change?**

No

### CODE REVIEW

File: lib/ContextModule.js [hunk 1]
```diff
@@ -2,218 +2,257 @@
...
+			.filter(dependency => dependency.module)
+			.sort((a, b) => {
+				if(a.userRequest === b.userRequest) {
+					return 0;
+				}
+				return a.userRequest < b.userRequest ? -1 : 1;
+			}).reduce(function(map, dep) {
+				map[dep.userRequest] = dep.module.id;
+				return map;
+			}, {});
```
Comment (sokra) [inline lib/ContextModule.js]: `{}` -> `Object.create(null)`

File: lib/ContextModule.js [hunk 2]
```diff
@@ -2,218 +2,257 @@
...
+
+	source() {
+		return this.getSource(
+			this.getSourceString()
+		);
+	}
+
+	size() {
+		// base penalty
+		let count = 160;
```
Comment (sokra) [inline lib/ContextModule.js]: const count

File: lib/ContextModule.js [hunk 3]
```diff
@@ -2,218 +2,257 @@
...
-	if(!this.recursive)
-		identifier += "nonrecursive ";
-	if(this.addon)
-		identifier += this.addon;
-	if(this.regExp)
-		identifier += this.regExp;
-	return identifier.replace(/ $/, "");
-};
+class ContextModule extends Module {
+	constructor(resolveDependencies, context, recursive, regExp, addon, async) {
```
Comment (Kovensky) [inline lib/ContextModule.js]: Don't use `async` as a bare identifier; maybe name it `isAsync` instead?

File: lib/ContextModule.js [hunk 4]
```diff
@@ -2,218 +2,257 @@
...
-	if(!this.recursive)
-		identifier += "nonrecursive ";
-	if(this.addon)
-		identifier += requestShortener.shorten(this.addon);
-	if(this.regExp)
-		identifier += prettyRegExp(this.regExp + "");
-	return identifier.replace(/ $/, "");
-};
+	contextify(context, request) {
+		return request.split("!").map(function(subrequest) {
```
Comment (Kovensky) [inline lib/ContextModule.js]: arrow function

File: lib/ContextModule.js [hunk 5]
```diff
@@ -2,218 +2,257 @@
...
+			}).reduce(function(map, item) {
+				const chunks = item.block.chunks || [];
+				if(chunks.length !== 1) {
+					hasMultipleOrNoChunks = true;
+				}
+				map[item.userRequest] = [item.dependency.module.id]
+					.concat(chunks.map(chunk => chunk.id));
+
+				return map;
+			}, {});
```
Comment (Kovensky) [inline lib/ContextModule.js]: These should be arrow functions, and the empty object `Object.create(null)`.

File: lib/ContextModule.js (hunk 2 above)
Comment (Kovensky) [inline lib/ContextModule.js]: Maybe `initialCount`? Avoids the apparently confusing shadowing in the reducer.

**Comment by timse** [comment]
thanks for the review :) @sokra & @Kovensky

**Comment by sokra** [comment]
Thanks

**Pull request merged**